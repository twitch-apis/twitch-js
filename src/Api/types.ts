import { ApiVersions } from '../twitch'

import { FetchOptions } from '../utils/fetch'
import { LoggerOptions } from '../utils/logger'

export type ApiOptions = {
  token: string
  clientId?: string
  log?: LoggerOptions
  onAuthenticationFailure?: () => Promise<string>
}

export enum ApiReadyStates {
  'NOT_READY',
  'READY',
  'INITIALIZED',
}

export type ApiSettings = {
  baseUrl: string
  authorizationHeader: string
}

export const Settings: Record<ApiVersions, ApiSettings> = {
  [ApiVersions.Helix]: {
    baseUrl: 'https://api.twitch.tv/helix',
    authorizationHeader: 'Bearer',
  },
  [ApiVersions.Kraken]: {
    baseUrl: 'https://api.twitch.tv/kraken',
    authorizationHeader: 'OAuth',
  },
}

export type ApiHeaders = Record<string, string> & {
  'Client-ID'?: string
  Accept?: string
  Authorization?: string
}

type ApiVersionOptions = {
  version?: ApiVersions
}

export type ApiFetchOptions = FetchOptions & ApiVersionOptions
