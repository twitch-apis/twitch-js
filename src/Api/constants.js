/**
 * @module Api
 */

export const HELIX_VERSION = 'helix'
export const HELIX_URL_ROOT = 'https://api.twitch.tv/helix'
export const HELIX_AUTHORIZATION_HEADER = 'Bearer'

export const KRAKEN_VERSION = 'kraken'
export const KRAKEN_URL_ROOT = 'https://api.twitch.tv/kraken'
export const KRAKEN_AUTHORIZATION_HEADER = 'OAuth'

/** @typedef {string} ApiReadyState */
/**
 * API client ready state
 * @readonly
 * @enum {ApiReadyState}
 * @property {string} 0 not ready
 * @property {string} 1 ready
 * @property {string} 2 initialized
 */
export const READY_STATES = {
  0: 'NOT_READY',
  1: 'READY',
  2: 'INITIALIZED',
}
