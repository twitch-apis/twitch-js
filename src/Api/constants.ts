import { ApiSettings } from './types'

export const Settings: Record<'Helix' | 'Kraken', ApiSettings> = {
  Helix: {
    BaseUrl: 'https://api.twitch.tv/helix',
    Bearer: 'Bearer',
  },
  Kraken: {
    BaseUrl: 'https://api.twitch.tv/kraken',
    Bearer: 'OAuth',
  },
}
