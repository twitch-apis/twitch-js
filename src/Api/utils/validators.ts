import invariant from 'invariant'

import conformsTo from 'lodash/conformsTo'
import defaults from 'lodash/defaults'
import isFunction from 'lodash/isFunction'
import isString from 'lodash/isString'
import isUndefined from 'lodash/isUndefined'

import { Options } from '../types'

export const apiOptions = (maybeOptions: any): Options | never => {
  const shape = {
    clientId: (token: unknown) => isUndefined(token) || isString(token),
    token: (token: unknown) => isUndefined(token) || isString(token),
    onAuthenticationFailure: isFunction,
  }

  const options = defaults<Options, Options>(
    { ...maybeOptions },
    {
      clientId: undefined,
      token: undefined,
      onAuthenticationFailure: () => new Promise((_, reject) => reject()),
    },
  )

  invariant(
    conformsTo(options, shape),
    '[twitch-js/Api] options: Expected valid options',
  )

  return options as Options
}
