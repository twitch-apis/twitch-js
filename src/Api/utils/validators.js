import invariant from 'invariant'
import { conformsTo, defaultsDeep, isBoolean, isString } from 'lodash'

import * as constants from '../constants'

const apiOptions = maybeOptions => {
  /**
   * API options
   * @typedef {Object} ApiOptions
   * @property {string} [clientId] Optional if token is defined.
   * @property {string} [token] Optional if clientId is defined.
   * @property {string} [urlRoot]
   * @property {boolean} [debug=false]
   */
  const shape = {
    debug: isBoolean,
  }

  const shapeWithClientId = {
    ...shape,
    clientId: isString,
  }

  const shapeWithToken = {
    ...shape,
    token: isString,
  }

  const options = defaultsDeep(
    {},
    { ...maybeOptions },
    {
      urlRoot: constants.KRAKEN_URL_ROOT,
      debug: false,
    },
  )

  invariant(
    conformsTo(options, shapeWithClientId) ||
      conformsTo(options, shapeWithToken),
    '[twitch-js/Api] options: Expected valid options',
  )

  return options
}

export { apiOptions }
