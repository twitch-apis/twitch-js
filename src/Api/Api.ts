import invariant from 'invariant'
import isEmpty from 'lodash/isEmpty'
import includes from 'lodash/includes'
import toLower from 'lodash/toLower'
import toUpper from 'lodash/toUpper'

import { ApiRootResponse, ApiVersions } from '../twitch'

import createLogger, { Logger } from '../utils/logger'

import fetchUtil from '../utils/fetch'
import * as Errors from '../utils/fetch/Errors'
import * as validators from './utils/validators'

import {
  ApiOptions,
  ApiReadyStates,
  ApiFetchOptions,
  Settings,
  ApiHeaders,
} from './types'

/**
 * Make requests to Twitch API.
 *
 * ## Initializing
 *
 * ```js
 * // With a token ...
 * const token = 'cfabdegwdoklmawdzdo98xt2fo512y'
 * const { api } = new TwitchJs({ token })
 *
 * // ... or with a client ID ...
 * const clientId = 'uo6dggojyb8d6soh92zknwmi5ej1q2'
 * const { api } = new TwitchJs({ clientId })
 * ```
 *
 * **Note:** The recommended way to initialize the API client is with a token.
 *
 * ## Making requests
 *
 * By default, the API client makes requests to the
 * [Helix API](https://dev.twitch.tv/docs/api), and exposes [[Api.get]],
 * [[Api.post]] and [[Api.put]] methods. Query string parameters and body
 * parameters are provided via `options.search` and `options.body` properties,
 * respectively.
 *
 * To make requests to the [Kraken/v5 API](https://dev.twitch.tv/docs/v5), use
 * `options.version = 'kraken'`
 *
 * ### Examples
 *
 * #### Get bits leaderboard
 * ```js
 * api
 *   .get('bits/leaderboard', { search: { user_id: '44322889' } })
 *   .then(response => {
 *     // Do stuff with response ...
 *   })
 * ```
 *
 * #### Get the latest Overwatch live streams
 * ```
 * api
 *   .get('streams', { version: 'kraken', search: { game: 'Overwatch' } })
 *   .then(response => {
 *     // Do stuff with response ...
 *   })
 * ```
 *
 * #### Start a channel commercial
 * ```
 * const channelId = '44322889'
 * api
 *   .post(`channels/${channelId}/commercial`, {
 *     version: 'kraken',
 *     body: { length: 30 },
 *   })
 *   .then(response => {
 *     // Do stuff with response ...
 *   })
 * ```
 */

class Api {
  private _options!: ApiOptions
  private _log: Logger

  private _readyState: ApiReadyStates = ApiReadyStates.READY

  private _status!: ApiRootResponse

  constructor(maybeOptions: ApiOptions = {}) {
    this.options = maybeOptions

    this._log = createLogger({ name: 'Api', ...this.options.log })
  }

  set options(maybeOptions) {
    this._options = validators.apiOptions(maybeOptions)
  }

  get options() {
    return this._options
  }

  get readyState() {
    return this._readyState
  }

  get status() {
    return this._status
  }

  /**
   * New client options. To update `token` or `clientId`, use [**api.initialize()**]{@link Api#initialize}.
   */
  updateOptions(options: Partial<ApiOptions>) {
    const { clientId, token } = this.options
    this.options = { ...options, clientId, token }
  }

  /**
   * Initialize API client and retrieve status.
   * @see https://dev.twitch.tv/docs/v5/#root-url
   */
  async initialize(newOptions?: Partial<ApiOptions>) {
    if (newOptions) {
      this.options = { ...this.options, ...newOptions }
    }

    if (!newOptions && this.readyState === 2) {
      return Promise.resolve()
    }

    const statusResponse = await this.get<ApiRootResponse>('', {
      version: ApiVersions.Kraken,
    })

    this._readyState = ApiReadyStates.INITIALIZED
    this._status = statusResponse

    return statusResponse
  }

  /**
   * Check if current credentials include `scope`.
   * @see https://dev.twitch.tv/docs/authentication/#twitch-api-v5
   */
  hasScope(
    /** Scope to check */
    scope: string,
  ): Promise<boolean> {
    return new Promise((resolve, reject) => {
      if (this.readyState !== 2 || !this.status) {
        return reject(false)
      }

      return includes(this.status.token.authorization.scopes, scope)
        ? resolve(true)
        : reject(false)
    })
  }

  /**
   * GET endpoint.
   *
   * @example <caption>Get Live Overwatch Streams (Kraken)</caption>
   * ```
   * api.get('streams', { version: 'kraken', search: { game: 'Overwatch' } })
   *   .then(response => {
   *     // Do stuff with response ...
   *   })
   * ```
   *
   * @example <caption>Get user follows (Helix)</caption>
   * ```
   * api.get('users/follows', { search: { to_id: '23161357' } })
   *   .then(response => {
   *     // Do stuff with response ...
   *   })
   * ```
   */
  get<T = any>(endpoint = '', options?: ApiFetchOptions) {
    return this._handleFetch<T>(endpoint, options)
  }

  /**
   * POST endpoint.
   */
  post<T = any>(endpoint: string, options?: ApiFetchOptions) {
    return this._handleFetch<T>(endpoint, { ...options, method: 'post' })
  }

  /**
   * PUT endpoint.
   */
  put<T = any>(endpoint: string, options?: ApiFetchOptions) {
    return this._handleFetch<T>(endpoint, { ...options, method: 'put' })
  }

  private _isVersionHelix(version: ApiVersions) {
    return toLower(version) === ApiVersions.Helix
  }

  private _getBaseUrl(version: ApiVersions) {
    return Settings[version].baseUrl
  }

  private _getHeaders(version: ApiVersions): ApiHeaders {
    const { clientId, token } = this.options

    const isHelix = this._isVersionHelix(version)

    invariant(
      isHelix ? !(isEmpty(clientId) && isEmpty(token)) : true,
      '[twitch-js/Api] To call a Helix endpoint, a `clientId` or `token` must be provided',
    )

    const headers: ApiHeaders = {}

    if (!isHelix) {
      headers['Accept'] = 'application/vnd.twitchtv.v5+json'
    }

    if (clientId) {
      headers['Client-ID'] = clientId
    }

    if (token) {
      headers[
        'Authorization'
      ] = `${Settings[version].authorizationHeader} ${token}`
    }

    return headers
  }

  private async _handleFetch<T = any>(
    maybeUrl = '',
    options: ApiFetchOptions = {},
  ) {
    const { version = ApiVersions.Helix, ...fetchOptions } = options

    const baseUrl = this._getBaseUrl(version)

    const url = `${baseUrl}/${maybeUrl}`

    const message = `${toUpper(fetchOptions.method) || 'GET'} ${url}`

    const fetchProfiler = this._log.profile()

    const headers = this._getHeaders(version)

    const optionHeaders = Object.entries(fetchOptions.headers || {})

    for (const [name, value] of optionHeaders) {
      headers[String(name)] = value
    }

    const performRequest = () =>
      fetchUtil<T>(url, {
        ...fetchOptions,
        headers,
      })

    try {
      return await performRequest()
    } catch (error) {
      if (
        error instanceof Errors.AuthenticationError &&
        typeof this.options.onAuthenticationFailure === 'function'
      ) {
        const token = await this.options.onAuthenticationFailure()

        if (token) {
          await this.initialize({ token })
          this._log.info(`${message} ... re-initializing with new token`)
        }

        this._log.info(`${message} ... retrying`)

        return await performRequest()
      }
      throw new Errors.FetchError(error, message)
    } finally {
      fetchProfiler.done(message)
    }
  }
}

export default Api
