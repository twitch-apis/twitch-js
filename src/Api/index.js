import { get, includes, pickBy, toUpper } from 'lodash'

import createLogger from '../utils/logger/create'

import fetchUtil from '../utils/fetch'
import * as Errors from '../utils/fetch/Errors'
import * as validators from './utils/validators'

import * as constants from './constants'

/**
 * API client
 *
 * @example <caption>Get Featured Streams</caption>
 * const token = 'cfabdegwdoklmawdzdo98xt2fo512y'
 * const username = 'ronni'
 * const { api } = new TwitchJs({ token, username })
 *
 * api.get('streams/featured').then(response => {
 *   // Do stuff ...
 * })
 */
class Api {
  /**
   * API constructor.
   * @param {ApiOptions} options
   */
  constructor(maybeOptions = {}) {
    this.options = maybeOptions

    this.log = createLogger({ scope: 'Api', ...this.options.log })

    /**
     * API ready state
     * @private
     * @type {ApiReadyState}
     */
    this._readyState = 1

    /**
     * API status state.
     * @typedef {Object} ApiStatusState
     * @property {Object} token
     * @property {Object} token.authorization
     * @property {Array<string>} token.authorization.scopes
     * @property {string} token.authorization.createdAt
     * @property {string} token.authorization.updatedAt
     * @property {string} token.clientId
     * @property {string} token.userId
     * @property {string} token.userName
     * @property {boolean} token.valid
     */
    /**
     * API status.
     * @private
     * @type {ApiStatusState}
     */
    this._status = {}
  }

  set options(maybeOptions) {
    this._options = validators.apiOptions(maybeOptions)
  }

  get options() {
    return this._options
  }

  get readyState() {
    return this._readyState
  }

  get status() {
    return this._status
  }

  /**
   * Update client options.
   * @param {ApiOptions} options New client options. To update `token` or `clientId`, use [**api.initialize()**]{@link Api#initialize}.
   */
  updateOptions(options) {
    const { clientId, token } = this.options
    this.options = { ...options, clientId, token }
  }

  getAuthorizationType(version) {
    if (toUpper(version) === 'HELIX') {
      return 'Bearer'
    }

    return 'OAuth'
  }

  getBaseUrl(version) {
    if (toUpper(version) === 'HELIX') {
      return constants.HELIX_URL_ROOT
    }

    return constants.KRAKEN_URL_ROOT
  }

  getHeaders(version) {
    const { clientId, token } = this.options
    const authType = this.getAuthorizationType(version)

    return pickBy({
      Accept: 'application/vnd.twitchtv.v5+json',
      'Client-ID': clientId ? clientId : undefined,
      Authorization: token ? `${authType} ${token}` : undefined,
    })
  }

  /**
   * Initialize API client and retrieve status.
   * @param {ApiOptions} [options] Provide new options to client.
   * @returns {Promise<ApiStatusState, Object>}
   * @see https://dev.twitch.tv/docs/v5/#root-url
   */
  initialize(newOptions) {
    if (newOptions) {
      this.options = { ...this.options, ...newOptions }
    }

    if (!newOptions && this.readyState === 2) {
      return Promise.resolve()
    }

    return this.get().then(statusResponse => {
      this._readyState = 2
      this._status = statusResponse

      return statusResponse
    })
  }

  /**
   * Check if current credentials include `scope`.
   * @param {string} scope Scope to check.
   * @return {Promise<boolean, boolean>}
   * @see https://dev.twitch.tv/docs/authentication/#twitch-api-v5
   */
  hasScope(scope) {
    return new Promise((resolve, reject) => {
      if (this.readyState !== 2) {
        return reject(false)
      }

      return includes(get(this.status, 'token.authorization.scopes', []), scope)
        ? resolve(true)
        : reject(false)
    })
  }

  /**
   * GET endpoint.
   * @param {string} endpoint
   * @param {FetchOptions} [options]
   *
   * @example <caption>Get Live Overwatch Streams</caption>
   * api.get('streams', { search: { game: 'Overwatch' } })
   *   .then(response => {
   *     // Do stuff with response ...
   *   })
   */
  get(endpoint, options = {}) {
    return handleFetch.call(this, endpoint, options)
  }

  /**
   * POST endpoint.
   * @param {string} endpoint
   * @param {FetchOptions} [options={method:'post'}]
   */
  post(endpoint, options = {}) {
    return handleFetch.call(this, endpoint, { ...options, method: 'post' })
  }

  /**
   * PUT endpoint.
   * @param {string} endpoint
   * @param {FetchOptions} [options={method:'put'}]
   */
  put(endpoint, options = {}) {
    return handleFetch.call(this, endpoint, { ...options, method: 'put' })
  }
}

function handleFetch(maybeUrl = '', options = {}) {
  const fetchProfiler = this.log.startTimer()

  const [, version, url] = /^(?:([a-z]+):)?\/?(.*)$/i.exec(maybeUrl)
  const baseUrl = `${this.getBaseUrl(version)}/${url}`

  const message = `${options.method || 'GET'} ${baseUrl}`

  const request = () =>
    fetchUtil(baseUrl, {
      ...options,
      headers: {
        ...options.headers,
        ...this.getHeaders(version),
      },
    }).then(res => {
      fetchProfiler.done({ message })
      return res
    })

  return request().catch(error => {
    fetchProfiler.done({ level: 'error', message: error.body })

    if (error instanceof Errors.AuthenticationError) {
      return this.options
        .onAuthenticationFailure()
        .then(token => (this.options = { ...this.options, token }))
        .then(() => this.log.info('Retrying (with new credentials)'))
        .then(() => request())
    }

    throw error
  })
}

export default Api
