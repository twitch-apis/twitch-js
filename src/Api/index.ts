import includes from 'lodash/includes'
import toLower from 'lodash/toLower'
import toUpper from 'lodash/toUpper'

import * as twitchTypes from '../twitch'

import createLogger, { Logger } from '../utils/logger'

import fetchUtil from '../utils/fetch'
import * as Errors from '../utils/fetch/Errors'
import * as validators from './utils/validators'

import * as types from './types'
export * from './types'

/**
 * @class
 * @public
 *
 * @example <caption>Get Featured Streams</caption>
 * ```
 * const token = 'cfabdegwdoklmawdzdo98xt2fo512y'
 * const username = 'ronni'
 * const { api } = new TwitchJs({ token, username })
 *
 * api.get('streams/featured').then(response => {
 *   // Do stuff ...
 * })
 * ```
 */
class Api {
  private _options: types.Options
  private _log: Logger

  private _readyState = 1

  private _status: twitchTypes.ApiRootResponse

  constructor(maybeOptions: types.Options) {
    /**
     * @type {ApiOptions}
     * @private
     */
    this.options = maybeOptions

    this._log = createLogger({ name: 'Api', ...this.options.log })
  }

  set options(maybeOptions) {
    this._options = validators.apiOptions(maybeOptions)
  }

  get options() {
    return this._options
  }

  get readyState() {
    return this._readyState
  }

  get status() {
    return this._status
  }

  /**
   * New client options. To update `token` or `clientId`, use [**api.initialize()**]{@link Api#initialize}.
   */
  updateOptions(options: Partial<types.Options>) {
    const { clientId, token } = this.options
    this.options = { ...options, clientId, token }
  }

  /**
   * Initialize API client and retrieve status.
   * @see https://dev.twitch.tv/docs/v5/#root-url
   */
  async initialize(newOptions?: Partial<types.Options>) {
    if (newOptions) {
      this.options = { ...this.options, ...newOptions }
    }

    if (!newOptions && this.readyState === 2) {
      return Promise.resolve()
    }

    const statusResponse = await this.get<twitchTypes.ApiRootResponse>()

    this._readyState = 2
    this._status = statusResponse

    return statusResponse
  }

  /**
   * Check if current credentials include `scope`.
   * @see https://dev.twitch.tv/docs/authentication/#twitch-api-v5
   */
  hasScope(
    /** Scope to check */
    scope: string,
  ): Promise<boolean> {
    return new Promise((resolve, reject) => {
      if (this.readyState !== 2 || !this.status) {
        return reject(false)
      }

      return includes(this.status.token.authorization.scopes, scope)
        ? resolve(true)
        : reject(false)
    })
  }

  /**
   * GET endpoint.
   *
   * @example <caption>Get Live Overwatch Streams (Kraken)</caption>
   * ```
   * api.get('streams', { version: 'kraken', search: { game: 'Overwatch' } })
   *   .then(response => {
   *     // Do stuff with response ...
   *   })
   * ```
   *
   * @example <caption>Get user follows (Helix)</caption>
   * ```
   * api.get('users/follows', { search: { to_id: '23161357' } })
   *   .then(response => {
   *     // Do stuff with response ...
   *   })
   * ```
   */
  get<T = any>(endpoint = '', options?: types.FetchOptions) {
    return this._handleFetch<T>(endpoint, options)
  }

  /**
   * POST endpoint.
   */
  post<T = any>(endpoint: string, options?: types.FetchOptions) {
    return this._handleFetch<T>(endpoint, { ...options, method: 'post' })
  }

  /**
   * PUT endpoint.
   */
  put<T = any>(endpoint: string, options?: types.FetchOptions) {
    return this._handleFetch<T>(endpoint, { ...options, method: 'put' })
  }

  private _isVersionHelix(version: twitchTypes.ApiVersions) {
    return toLower(version) === twitchTypes.ApiVersions.Helix
  }

  private _getBaseUrl(version: twitchTypes.ApiVersions) {
    return types.Settings[version].baseUrl
  }

  private _getHeaders(version: twitchTypes.ApiVersions): types.Headers {
    const { clientId, token } = this.options

    const isHelix = this._isVersionHelix(version)

    const headers = isHelix
      ? { 'Client-ID': clientId }
      : { Accept: 'application/vnd.twitchtv.v5+json', 'Client-ID': clientId }

    if (token) {
      const authorizationHeader = types.Settings[version].authorizationHeader
      const authorization = `${authorizationHeader} ${token}`

      return { ...headers, Authorization: authorization }
    }

    return headers
  }

  private async _handleFetch<T = any>(
    maybeUrl = '',
    options: types.FetchOptions = {},
  ) {
    const { version = twitchTypes.ApiVersions.Helix, ...fetchOptions } = options

    const baseUrl = this._getBaseUrl(version)

    const url = `${baseUrl}/${maybeUrl}`

    const message = `${toUpper(fetchOptions.method) || 'GET'} ${url}`

    const fetchProfiler = this._log.startTimer(message)

    const performRequest = () =>
      fetchUtil<T>(url, {
        ...fetchOptions,
        headers: {
          ...fetchOptions.headers,
          ...this._getHeaders(version),
        },
      })

    try {
      return await performRequest()
    } catch (error) {
      if (error instanceof Errors.AuthenticationError) {
        const token = await this.options.onAuthenticationFailure()

        if (token) {
          await this.initialize({ token })
          this._log.info(`${message} ... retrying with new token`)
        } else {
          this._log.info(`${message} ... retrying`)
        }

        return await performRequest()
      }
      throw new Errors.FetchError(error, message)
    } finally {
      fetchProfiler.done(message)
    }
  }
}

export default Api
