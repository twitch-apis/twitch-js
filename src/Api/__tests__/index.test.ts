import { mocked } from 'ts-jest/utils'
import nodeFetch from 'node-fetch'

import * as twitchTypes from '../../twitch'

import * as Errors from '../../utils/fetch/Errors'

import Api from '../'
import { Settings } from '../types'

jest.mock('node-fetch')

const fetchUtil = mocked(nodeFetch, true)

describe('Api', () => {
  jest.setTimeout(500)

  const options = {
    token: 'TOKEN',
    clientId: 'CLIENT_ID',
    log: { enabled: false },
  }

  const fetchOptions = { search: { a: 'b' } }

  const helixBaseUrl = Settings.helix.baseUrl
  const krakenBaseUrl = Settings.kraken.baseUrl

  afterEach(() => {
    fetchUtil.mockClear()
  })

  describe('constructor', () => {
    test('should instantiate with no options', () => {
      expect(() => new Api({ log: { enabled: false } })).not.toThrow()
    })
  })

  test('should allow options to be updated without changing clientId or token', () => {
    const api = new Api(options)

    const log = { level: 'info' }

    const nextOptions = {
      clientId: 'NEXT_CLIENT_ID',
      token: 'NEXT_TOKEN',
      log,
    }

    api.updateOptions(nextOptions)

    expect(api.options).toMatchObject({ ...options, log })
  })

  describe('initialize', () => {
    test('should set readyState and status', async () => {
      const api = new Api(options)
      await api.initialize()

      expect(api.readyState).toBe(2)
      expect(api.status).toMatchSnapshot()
    })

    test('should resolve if already initialized', async () => {
      const api = new Api(options)
      await api.initialize()
      const actual = await api.initialize()

      expect(actual).toMatchSnapshot()
    })

    test('should update client options', async () => {
      const api = new Api(options)
      expect(api.options.token).toBe(options.token)

      const newToken = 'NEW_TOKEN'
      await api.initialize({ token: newToken })

      expect(api.options.token).toBe(newToken)
    })
  })

  describe('headers', () => {
    test('should create headers with clientId and token', async () => {
      const api = new Api(options)
      await api.get()

      expect(fetchUtil.mock.calls).toMatchSnapshot()
    })

    test('should create headers with clientId', async () => {
      const api = new Api({ ...options, token: undefined })
      await api.get()

      expect(fetchUtil.mock.calls).toMatchSnapshot()
    })

    test('should create headers for Kraken', async () => {
      const api = new Api(options)
      await api.get('', { version: twitchTypes.ApiVersions.Kraken })

      expect(fetchUtil.mock.calls).toMatchSnapshot()
    })

    test('should create headers for Helix', async () => {
      const api = new Api(options)
      await api.get('', { version: twitchTypes.ApiVersions.Helix })

      expect(fetchUtil.mock.calls).toMatchSnapshot()
    })

    test('should throw if clientId and token are absent when calling a Helix endpoint', async () => {
      const api = new Api({ log: { enabled: false } })

      await expect(
        api.get('', { version: twitchTypes.ApiVersions.Helix }),
      ).rejects.toMatchInlineSnapshot(
        `[Invariant Violation: [twitch-js/Api] To call a Helix endpoint, a \`clientId\` or \`token\` must be provided]`,
      )
    })
  })

  describe('hasScope', () => {
    test('should reject if instance is uninitialized', async () => {
      const api = new Api(options)
      await expect(api.hasScope('user_read')).rejects.toBe(false)
    })

    test('should reject if scope is absent', async () => {
      const api = new Api(options)
      await expect(api.hasScope('channel_read')).rejects.toBe(false)
    })

    test('should resolve if scope is present', async () => {
      const api = new Api(options)
      await api.initialize()

      await expect(api.hasScope('user_read')).resolves.toBe(true)
    })
  })

  describe('methods', () => {
    test('get should call fetch', async () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'
      await api.get(endpoint, fetchOptions)

      const [actualEndpoint, actualOpts] = fetchUtil.mock.calls[0]

      expect(actualEndpoint).toEqual(expect.stringContaining(endpoint))
      expect(actualOpts).toMatchObject(fetchOptions)
    })

    test('post should call fetch with method=post', async () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'
      await api.post(endpoint, fetchOptions)

      const [, actualOpts] = fetchUtil.mock.calls[0]

      expect(actualOpts).toMatchObject({ method: 'post' })
    })

    test('put should call fetch with method=put', async () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'
      await api.put(endpoint, fetchOptions)

      const [, actualOpts] = fetchUtil.mock.calls[0]

      expect(actualOpts).toMatchObject({ method: 'put' })
    })

    test('should throw on failure', async () => {
      const api = new Api(options)

      try {
        await api.get('404')
      } catch (error) {
        expect(error).toBeInstanceOf(Errors.FetchError)
        expect(error).toMatchSnapshot()
      }
    })
  })

  describe('versions', () => {
    test('should fallback to the Helix endpoint', () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'

      return api.get(endpoint, fetchOptions).then(() => {
        const [actualEndpoint, actualOpts] = fetchUtil.mock.calls[0]

        expect(actualEndpoint).toBe(`${helixBaseUrl}/${endpoint}?a=b`)
        expect(actualOpts).toMatchObject(fetchOptions)
      })
    })

    test('should call the Kraken endpoint', () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'

      return api
        .get(endpoint, {
          ...fetchOptions,
          version: twitchTypes.ApiVersions.Kraken,
        })
        .then(() => {
          const [actualEndpoint, actualOpts] = fetchUtil.mock.calls[0]

          expect(actualEndpoint).toBe(`${krakenBaseUrl}/${endpoint}?a=b`)
          expect(actualOpts).toMatchObject(fetchOptions)
        })
    })

    test('should call the Helix endpoint', () => {
      const api = new Api(options)

      const endpoint = 'ENDPOINT'
      const opts = {
        ...fetchOptions,
        version: twitchTypes.ApiVersions['Helix'],
      }

      return api.get(endpoint, opts).then(() => {
        const [actualEndpoint, actualOpts] = fetchUtil.mock.calls[0]

        expect(actualEndpoint).toBe(`${helixBaseUrl}/${endpoint}?a=b`)
        expect(actualOpts).toMatchSnapshot()
      })
    })
  })

  describe('onAuthenticationFailure', () => {
    test('should call onAuthenticationFailure', async () => {
      const onAuthenticationFailure = jest.fn(() => Promise.reject())
      const api = new Api({
        ...options,
        token: 'INVALID_TOKEN',
        onAuthenticationFailure,
      })

      try {
        await api.get('401')
      } catch (error) {
        expect(onAuthenticationFailure).toHaveBeenCalled()
      }
    })

    test('should update token', async () => {
      const onAuthenticationFailure = jest.fn(() => Promise.resolve('TOKEN'))
      const api = new Api({
        ...options,
        token: 'INVALID_TOKEN',
        onAuthenticationFailure,
      })

      try {
        await api.get('401')
      } catch (error) {
        expect(onAuthenticationFailure).toHaveBeenCalled()
        expect(api.options.token).toEqual('TOKEN')
      }
    })
  })
})
