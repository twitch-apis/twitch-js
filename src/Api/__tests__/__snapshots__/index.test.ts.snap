// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api headers should create headers for Helix 1`] = `
Array [
  Array [
    "https://api.twitch.tv/helix/",
    Object {
      "headers": Object {
        "Authorization": "Bearer TOKEN",
        "Client-ID": "CLIENT_ID",
      },
      "method": "get",
    },
  ],
]
`;

exports[`Api headers should create headers for Kraken 1`] = `
Array [
  Array [
    "https://api.twitch.tv/kraken/",
    Object {
      "headers": Object {
        "Accept": "application/vnd.twitchtv.v5+json",
        "Authorization": "OAuth TOKEN",
        "Client-ID": "CLIENT_ID",
      },
      "method": "get",
    },
  ],
]
`;

exports[`Api headers should create headers with clientId 1`] = `
Array [
  Array [
    "https://api.twitch.tv/helix/",
    Object {
      "headers": Object {
        "Client-ID": "CLIENT_ID",
      },
      "method": "get",
    },
  ],
]
`;

exports[`Api headers should create headers with clientId and token 1`] = `
Array [
  Array [
    "https://api.twitch.tv/helix/",
    Object {
      "headers": Object {
        "Authorization": "Bearer TOKEN",
        "Client-ID": "CLIENT_ID",
      },
      "method": "get",
    },
  ],
]
`;

exports[`Api initialize should resolve if already initialized 1`] = `undefined`;

exports[`Api initialize should set readyState and status 1`] = `
Object {
  "token": Object {
    "authorization": Object {
      "createdAt": "2016-12-14T15:51:16Z",
      "scopes": Array [
        "user_read",
      ],
      "updatedAt": "2016-12-14T15:51:16Z",
    },
    "clientId": "uo6dggojyb8d6soh92zknwmi5ej1q2",
    "userId": "44322889",
    "userName": "dallas",
    "valid": true,
  },
}
`;

exports[`Api methods should throw on failure 1`] = `[Error: [TwitchJS] https://api.twitch.tv/helix/404 Not Found]`;

exports[`Api versions should call the Helix endpoint 1`] = `
Object {
  "headers": Object {
    "Authorization": "Bearer TOKEN",
    "Client-ID": "CLIENT_ID",
  },
  "method": "get",
  "search": Object {
    "a": "b",
  },
}
`;
